src/coreclr/jit/helperexpansion.cpp

```diff
@@ -642,6 +654,17 @@ bool Compiler::fgExpandThreadLocalAccessForCall(BasicBlock** pBlock, Statement*
         fgNewBBFromTreeAfter(BBJ_ALWAYS, threadStaticBlockNullCondBB, fallbackValueDef, debugInfo, true);
 
     // fastPathBb
+    if (isGCThreadStatic)
+    {
+        // Need to add extra indirection to access the data pointer.
+
+        threadStaticBlockBaseLclValueUse =
+            gtNewIndir(TYP_I_IMPL, threadStaticBlockBaseLclValueUse, GTF_IND_NONFAULTING);
```

Comment:
```suggestion
            gtNewIndir(TYP_REF, threadStaticBlockBaseLclValueUse, GTF_IND_NONFAULTING);
```
This is object reference. Needs to be tracked by the GC.

Reply:
hhm, I thought the `base + dataOffset` is the only thing we need to track, but the base value is unchanged and hence we are storing it in cache, no?

Reply:
The pointer to the handle does not need to be tracked. It is what we are storing in the cache.

The value that you get by dereferencing the handle needs to be tracked.

 If tracking of this value was not needed, you would need to track the value that you get by adding `threadStaticBlocksInfo.offsetOfGCDataPointer` to it - that you have found by trial and error that it is really needed.

Reply:
ah yes. we are doing `GT_IND(handle)` here which needs tracking. Makes sense.